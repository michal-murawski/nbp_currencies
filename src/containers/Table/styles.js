export const TableHeadStyles = theme => ({
  title: {
    padding: `${theme.gutters[32]} ${theme.gutters[40]}`,
    color: theme.palette.common.white,  
  },
});
