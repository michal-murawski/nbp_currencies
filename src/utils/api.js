import { generateRandomNumber } from 'utils/dataHelpers';
import options from './api.options';

const apiUrl = 'http://api.nbp.pl/api/exchangerates/tables/a';
const wrongApiUrl = 'http://api.nbp.pl/api/exchangerates/tables/f';
const localDevServer = 'http://localhost:7878/favourites';

async function badRequestWithError() {
  const response = await fetch(wrongApiUrl, options.get);

  if (!response.ok) {
    throw new Error('We could not download currencies!');
  }

  return await response.json();
}

async function getCurrencies() {
  const response = await fetch(apiUrl, options.get);

  if (!response.ok) {
    throw new Error('We could not download currencies!');
  }

  return await response.json();
}

async function getFavourites() {
  const response = await fetch(localDevServer, options.get);

  if (!response.ok) {
    throw new Error('We could not download favourites!');
  }

  return await response.json();
}

async function addFavourites(currency) {
  const body = JSON.stringify({
    id: generateRandomNumber(),
    code: currency,
  });

  try {
    const response = await fetch(localDevServer, options.post({ body }));

    return await response.json();
  } catch (e) {
    throw new Error('We could add favourite currency!');
  }
}

async function deleteFavourites(id) {
  const response = await fetch(`${localDevServer}/${id}`, options.delete);

  if (!response.ok) {
    throw new Error('We could not delete currency!');
  }

  return await response.json();
}

export default {
  currencies: {
    getCurrencies,
  },
  favourites: {
    getFavourites,
    addFavourites,
    deleteFavourites,
  },
};
