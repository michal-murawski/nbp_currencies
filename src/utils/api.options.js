const getOptions = {
  headers: {
    Accept: 'application/json',
  },
  method: 'GET',
};
const deleteOptions = {
  method: 'DELETE',
};
const postOptions = extraOptions => ({
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
  method: 'POST',
  ...extraOptions,
});

export default {
  get: getOptions,
  delete: deleteOptions,
  post: postOptions,
};
