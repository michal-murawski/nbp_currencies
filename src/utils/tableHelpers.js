import { map, prop } from 'ramda';


export const getRowKeys = map(prop('key'));
