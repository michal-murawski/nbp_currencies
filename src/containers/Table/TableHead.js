import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import {
  TableHead as TableHeadMUI,
  TableRow,
  TableCell,
  TableSortLabel,
} from 'material-ui/Table';
import { TableHeadStyles } from './styles';

class TableHead extends Component {
  constructor() {
    super();

    this.state = {
      activeIndex: null,
    };
  }

  handleOnClick = (sorter, index) => () => {
    if (sorter) {
      this.setState({ activeIndex: index });
      this.props.onHeadCellClick(sorter);
    }
  };

  render() {
    const { classes, headerLabels, direction } = this.props;
    const { activeIndex } = this.state;

    return (
      <TableHeadMUI>
        <TableRow classes={{ root: classes.tableHeadRow }}>
          {headerLabels.map((label, index) => (
            <TableCell
              onClick={this.handleOnClick(label.sorter, index)}
              key={label.label}
              classes={{ head: classes.title }}
            >
              {label.sorter ? (
                <TableSortLabel direction={direction} active={activeIndex === index}>
                  {label.label}
                </TableSortLabel>
              ) : (
                label.label
              )}
            </TableCell>
          ))}
        </TableRow>
      </TableHeadMUI>
    );
  }
}

TableHead.propTypes = {
  onHeadCellClick: PropTypes.func,
  direction: PropTypes.oneOf(['asc', 'desc']),
  headerLabels: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      sorter: PropTypes.string,
    })
  ),
};

export default withStyles(TableHeadStyles)(TableHead);
