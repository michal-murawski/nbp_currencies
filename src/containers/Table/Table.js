import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TableMUI from 'material-ui/Table';
import Paper from 'material-ui/Paper';
import { sortArray } from 'utils/dataHelpers';
import { getRowKeys } from 'utils/tableHelpers';
import TableHead from './TableHead';
import TableBody from './TableBody';

class Table extends Component {
  constructor(props) {
    super(props);

    this.state = {
      rows: props.rows,
      sortDirection: 'asc',
      searchParameter: '',
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState(({ searchParameter, sortDirection, rows }) => ({
      rows: sortArray(searchParameter, sortDirection, nextProps.rows),
    }));
  }

  getSortDirection(sortDirection) {
    if (sortDirection === 'asc') {
      return 'desc';
    } else {
      return 'asc';
    }
  }

  sortTableBy = searchParameter => {
    this.setState(({ rows, sortDirection }) => ({
      rows: sortArray(searchParameter, this.getSortDirection(sortDirection), rows),
      sortDirection: this.getSortDirection(sortDirection),
      searchParameter,
    }));
  };

  render() {
    const { headerLabels } = this.props;
    const { rows, sortDirection } = this.state;

    return (
      <Paper>
        <TableMUI>
          <TableHead
            headerLabels={headerLabels}
            onHeadCellClick={this.sortTableBy}
            direction={sortDirection}
          />
          <TableBody rows={rows} rowKeys={getRowKeys(headerLabels)} />
        </TableMUI>
      </Paper>
    );
  }
}

Table.defaultProps = {
  rows: [],
};

Table.propTypes = {
  rows: PropTypes.array,
};

export default Table;
