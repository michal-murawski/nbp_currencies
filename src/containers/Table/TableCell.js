import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import { TableCell as TableCellMUI } from 'material-ui/Table';

const styles = theme => ({
  root: {
    textAlign: 'left',
    fontSize: 18
  }
});

function TableCell(props) {
  const { children, classes } = props;

  return (
    <TableCellMUI classes={{ root: classes.root }}>{children}</TableCellMUI>
  );
}

TableCell.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

export default withStyles(styles)(TableCell);
