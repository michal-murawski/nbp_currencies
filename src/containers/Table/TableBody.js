import React from 'react';
import PropTypes from 'prop-types';
import { propOr } from 'ramda';
import { TableBody as TableBodyMUI, TableRow } from 'material-ui/Table';

import TableCell from './TableCell';

const generateKey = () => Math.floor(Math.random() * 1000);

function TableBody(props) {
  const { rows, rowKeys } = props;

  return (
    <TableBodyMUI>
      {rows.map(row => (
        <TableRow key={propOr(generateKey(), 'code', row)}>
          {rowKeys.map(key => 
            <TableCell key={propOr(generateKey(), key, row)}>{propOr('---', key, row)}</TableCell>
          )}
        </TableRow>
      ))}
    </TableBodyMUI>
  );
}


TableBody.propTypes = {
  type: PropTypes.string,
  rows: PropTypes.array,
  rowKeys: PropTypes.array,
};

export default TableBody;
