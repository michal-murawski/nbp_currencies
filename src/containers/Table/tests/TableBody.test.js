import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { currencies, tableHeadLabels } from 'tests/testData';
import { TableRow } from 'material-ui/Table';

import { getRowKeys } from 'utils/tableHelpers';
import TableBody from '../TableBody';
import TableCell from '../TableCell';

let count = 0;
function mockRandomNumber() {
  return count++;
}

jest.mock('utils/dataHelpers', () => ({
  generateRandomNumber: jest.fn(mockRandomNumber),
}));

describe('containers/Table/TableBody', () => {
  it('should render proper number of cells', () => {
    const wrapper = shallow(
      <TableBody rows={currencies} rowKeys={getRowKeys(tableHeadLabels)} />
    );

    const rows = wrapper.find(TableRow);
    const cellsLength = Object.keys(currencies[0]).length;
    const favIconLength = 1;
    expect(rows).toHaveLength(currencies.length);

    rows.forEach(row => {
      expect(row.find(TableCell)).toHaveLength(cellsLength + favIconLength);
    });
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
