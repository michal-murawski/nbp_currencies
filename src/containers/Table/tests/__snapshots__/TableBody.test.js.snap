// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Table/TableBody should render proper number of cells 1`] = `
<WithStyles(TableBody)>
  <WithStyles(TableRow)
    key="THB"
  >
    <WithStyles(TableCell)
      key="bat (Tajlandia)"
    >
      bat (Tajlandia)
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="THB"
    >
      THB
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="0.1099"
    >
      0.1099
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="4"
    >
      ---
    </WithStyles(TableCell)>
  </WithStyles(TableRow)>
  <WithStyles(TableRow)
    key="USD"
  >
    <WithStyles(TableCell)
      key="dolar amerykański"
    >
      dolar amerykański
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="USD"
    >
      USD
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="3.4319"
    >
      3.4319
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="9"
    >
      ---
    </WithStyles(TableCell)>
  </WithStyles(TableRow)>
  <WithStyles(TableRow)
    key="AUD"
  >
    <WithStyles(TableCell)
      key="dolar australijski"
    >
      dolar australijski
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="AUD"
    >
      AUD
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="2.6472"
    >
      2.6472
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="14"
    >
      ---
    </WithStyles(TableCell)>
  </WithStyles(TableRow)>
  <WithStyles(TableRow)
    key="HKD"
  >
    <WithStyles(TableCell)
      key="dolar Hongkongu"
    >
      dolar Hongkongu
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="HKD"
    >
      HKD
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="0.4372"
    >
      0.4372
    </WithStyles(TableCell)>
    <WithStyles(TableCell)
      key="19"
    >
      ---
    </WithStyles(TableCell)>
  </WithStyles(TableRow)>
</WithStyles(TableBody)>
`;
